### ISSUE TITLE
Trying to access undefined foreign variable "rl_num_chars_to_read". Building on OSX with readline keg-only.

### ISSUE THREAD

**Issue by svetlyak40wt** [vindarel/replic#1]
I cloned a repo and runned `make build` as README suggests. But this does not work. Probably, it is because it depends on some custom SBCL config which you have and which configures quicklisp. Because I don't have any custom config in `~/.sbclrc`.

```
[art@art-osx5:...sp/local-projects/replic]% make build
sbcl	--non-interactive \
		--load replic.asd \
		--eval '(ql:quickload :replic)' \
		--eval '(asdf:make :replic)'
This is SBCL 1.4.10, an implementation of ANSI Common Lisp.
More information about SBCL is available at <http://www.sbcl.org/>.

SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.
WARNING:
   Invalid :version specifier 0.1 for component "replic" from file
   #P"/Users/art/quicklisp/local-projects/replic/replic.asd",
   Substituting a string
Unhandled SB-INT:SIMPLE-READER-PACKAGE-ERROR in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                                          {10005505B3}>:
  Package QL does not exist.

    Line: 1, Column: 12, File-Position: 12

    Stream: #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}>

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {10005505B3}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:SIMPLE-READER-PACKAGE-ERROR "Package ~A does not exist." {1003A52CA3}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:SIMPLE-READER-PACKAGE-ERROR "Package ~A does not exist." {1003A52CA3}>)
2: (INVOKE-DEBUGGER #<SB-INT:SIMPLE-READER-PACKAGE-ERROR "Package ~A does not exist." {1003A52CA3}>)
3: (ERROR SB-INT:SIMPLE-READER-PACKAGE-ERROR :PACKAGE "QL" :STREAM #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}> :FORMAT-CONTROL "Package ~A does not exist." :FORMAT-ARGUMENTS ("QL"))
4: (SB-IMPL::READER-FIND-PACKAGE "QL" #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}>)
5: (SB-IMPL::READ-TOKEN #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}> #\q)
6: (SB-IMPL::READ-MAYBE-NOTHING #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}> #\q)
7: (SB-IMPL::READ-LIST #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}> #<unused argument>)
8: (SB-IMPL::READ-MAYBE-NOTHING #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}> #\()
9: (SB-IMPL::%READ-PRESERVING-WHITESPACE #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}> T (NIL) T)
10: (SB-IMPL::%READ-PRESERVING-WHITESPACE #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}> T (NIL) NIL)
11: (READ #<SB-IMPL::STRING-INPUT-STREAM {1003A52C13}> T NIL NIL)
12: (SB-IMPL::%READ-FROM-STRING "(ql:quickload :replic)" T NIL 0 NIL NIL)
13: (SB-IMPL::PROCESS-EVAL/LOAD-OPTIONS ((:LOAD . "replic.asd") (:EVAL . "(ql:quickload :replic)") (:EVAL . "(asdf:make :replic)") (:QUIT)))
14: (SB-IMPL::TOPLEVEL-INIT)
15: ((FLET SB-UNIX::BODY :IN SAVE-LISP-AND-DIE))
16: ((FLET "WITHOUT-INTERRUPTS-BODY-36" :IN SAVE-LISP-AND-DIE))
17: ((LABELS SB-IMPL::RESTART-LISP :IN SAVE-LISP-AND-DIE))

unhandled condition in --disable-debugger mode, quitting
make: *** [build] Error 1
```

**Comment by vindarel** [comment]
indeed, I have the lines added by `(ql:add-to-init-file)`:

```
;;; The following lines added by ql:add-to-init-file:
#-quicklisp
(let ((quicklisp-init (merge-pathnames "quicklisp/setup.lisp"
                                       (user-homedir-pathname))))
  (when (probe-file quicklisp-init)
    (load quicklisp-init)))
```

without them I get a different error:

```
    Package ASDF does not exist.

      Line: 5, Column: 15, File-Position: 62

      Stream: #<SB-SYS:FD-STREAM
                for "file /home/vince/projets/replic/replic.asd" {10039D73B3}>
```

(SBCL 1.2.4 Debian, ASDF 3.3.1)

This make command also works on CI (with the daewok docker image).

Don't you want to add those lines to your .sbclr ?

**Comment by svetlyak40wt** [comment]
Don't like such customization because the same reason – it is easy to forget about them and a library or program will not load or build for others.

**Comment by svetlyak40wt** [comment]
I was able to build `replic` with qlot and roswell + SBCL and CCL, using these commands:

```
qlot exec ros run -L ccl-bin --eval '(ql:quickload :replic)' --eval '(asdf:make :replic)'
mv replic replic-ccl

qlot exec ros run -L sbcl --eval '(ql:quickload :replic)' --eval '(asdf:make :replic)'
mv replic replic-sbcl
```

But they are not start with different errors:

```
% ./replic-sbcl
Unknown error:
Trying to access undefined foreign variable "rl_num_chars_to_read".
```

````
% ./replic-ccl
The value (:INTERNAL CFFI-CALLBACKS::|#::TEMP37047| CL-READLINE:REGISTER-FUNCTION) is not of the expected type (AND SYMBOL (NOT (SATISFIES CONSTANTP))).
```

**Comment by vindarel** [comment]
damn. What is your Quicklisp dist version ? You're on OSx ? You might be missing libreadline-dev ?

**Comment by svetlyak40wt** [comment]
That is on OSX.

I have readline installed with brew, but it says that the library is not linked into the `/usr/local`:

```
readline: stable 8.0.0 [keg-only]
Library for command-line editing
https://tiswww.case.edu/php/chet/readline/rltop.html
/usr/local/Cellar/readline/6.3.8 (46 files, 2.0MB)
  Poured from bottle on 2015-09-11 at 17:22:41
/usr/local/Cellar/readline/7.0 (46 files, 2MB)
  Poured from bottle on 2016-10-23 at 17:11:07
/usr/local/Cellar/readline/7.0.3_1 (46 files, 1.5MB)
  Poured from bottle on 2017-04-15 at 09:27:49
/usr/local/Cellar/readline/7.0.5 (46 files, 1.5MB)
  Poured from bottle on 2018-09-01 at 08:56:46
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/readline.rb
==> Caveats
readline is keg-only, which means it was not symlinked into /usr/local,
because macOS provides the BSD libedit library, which shadows libreadline.
In order to prevent conflicts when programs look for libreadline we are
defaulting this GNU Readline installation to keg-only.
```

Don't know how to work it around.

Is there an easy way to check if lisp is able to find a `readline` library?

**Comment by svetlyak40wt** [comment]
Quicklisp is latest 2019-01-07, installed by qlot locally. Here is the qlfile.lock:

```
("quicklisp" .
 (:class qlot.source.ql:source-ql-all
  :initargs (:distribution "http://beta.quicklisp.org/dist/quicklisp.txt" :%version :latest)
  :version "2019-01-07"))
```

**Comment by vindarel** [comment]
Looking at how cl-readline searches the library:

```
(define-foreign-library readline
  (:darwin (:or "libreadline.dylib"))
  (:unix   (:or "libreadline.so.6.3"
                "libreadline.so.6"
                "libreadline.so"))
  (t       (:default "libreadline")))
```

there might be something to do here.

**Comment by svetlyak40wt** [comment]
Not sure if this will be helpful, but I was able to run these commands without errors:

```
CL-USER> (cffi:define-foreign-library readline
  (:darwin (:or "libreadline.dylib"))
  (:unix   (:or "libreadline.so.6.3"
                "libreadline.so.6"
                "libreadline.so"))
  (t       (:default "libreadline")))
READLINE
CL-USER> (cffi:use-foreign-library readline)
#<FOREIGN-LIBRARY READLINE "libreadline.dylib">
CL-USER>
```

**Comment by svetlyak40wt** [comment]
I've made another attempt to run replic.

I've found 4 versions of the readline on my machine:

* /Users/art/RootLibrary/Frameworks/R.framework/Versions/3.0/Resources/lib/libreadline.dylib
* /usr/local/Cellar/readline/8.0.0/lib/libreadline.dylib
* /usr/lib/libreadline.dylib
* /opt/vagrant/embedded/lib/libreadline.dylib

Some of them are symlinks:

```
[art@lispbook:~replic]% ll /Users/art/RootLibrary/Frameworks/R.framework/Versions/3.0/Resources/lib/libreadline.dylib
lrwxr-xr-x  1 art  593637566    21B Mar  5  2014 /Users/art/RootLibrary/Frameworks/R.framework/Versions/3.0/Resources/lib/libreadline.dylib -> libreadline.5.2.dylib
[art@lispbook:~replic]% ll /usr/local/Cellar/readline/8.0.0/lib/libreadline.dylib
lrwxr-xr-x  1 art  593637566    21B Jan 30 14:00 /usr/local/Cellar/readline/8.0.0/lib/libreadline.dylib -> libreadline.8.0.dylib
[art@lispbook:~replic]% ll /usr/lib/libreadline.dylib
lrwxr-xr-x  1 root  wheel    15B Apr  8 17:31 /usr/lib/libreadline.dylib -> libedit.3.dylib
[art@lispbook:~replic]% ll /opt/vagrant/embedded/lib/libreadline.dylib
-rwxr-xr-x  1 root  wheel   654K Jul 17  2015 /opt/vagrant/embedded/lib/libreadline.dylib
```

I tried to copy each of them to the current directory and replic worked like a charm with every version except `/usr/lib/libreadline.dylib`.

Probably you need to add some way to detect of `brew` version of a library is available and to load it instead of "broken" system version of libreadline.

**Comment by svetlyak40wt** [comment]
Here is additional info, provided by brew on OSX:

```
==> Caveats
readline is keg-only, which means it was not symlinked into /usr/local,
because macOS provides the BSD libedit library, which shadows libreadline.
In order to prevent conflicts when programs look for libreadline we are
defaulting this GNU Readline installation to keg-only.

For compilers to find readline you may need to set:
  export LDFLAGS="-L/usr/local/opt/readline/lib"
  export CPPFLAGS="-I/usr/local/opt/readline/include"

For pkg-config to find readline you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/readline/lib/pkgconfig"
```

**Comment by svetlyak40wt** [comment]
This issue could be fixed by these two solutions:

* https://github.com/vindarel/cl-readline/pull/13
* https://github.com/vindarel/cl-readline/pull/14

I've checked them both. Both works on my OSX without need to copy any libraries to the current directory.

**Comment by vindarel** [comment]
I merged #14. Thanks !
